File: 0101-0200/0129.Sum_Root_to_Leaf_Numbers.cpp
//0ms

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
   void find(TreeNode* root,long long rem,int& ans){
    if(!root){
        return;
    }
    if(!root->left && !root->right){
        rem=rem*10+root->val;
        ans+=rem;
    }
    rem=rem*10+root->val;
    find(root->left,rem,ans);
    find(root->right,rem,ans);
   }
    int sumNumbers(TreeNode* root) {
        int ans=0;
        long long rem=0;
        find(root,rem,ans);
        return ans;
      
        
    }
};

File: 0101-0200/0129.Sum_Root_to_Leaf_Numbers.rs
//0ms


// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, mut sum: i32) -> i32 {
            if let Some(node_rc) = node {
                let node_ref = node_rc.borrow();
                sum = (sum * 10) + node_ref.val;
                if node_ref.left.is_none() && node_ref.right.is_none() {
                    sum
                } else {
                    dfs(&node_ref.left, sum) + dfs(&node_ref.right, sum)
                }
            } else {
                0
            }
        }
        dfs(&root, 0)
    }
}
